store/sqlstore/querybuilder/query.go

```diff
@@ -0,0 +1,257 @@
+	clone := q.clone()
+
+	clone.whereStatements = make([]string, len(q.whereStatements)+1)
+	copy(clone.whereStatements, q.whereStatements)
+	clone.whereStatements[len(q.whereStatements)] = sql
+
+	return clone
+}
+
+// OrderBy clones the query object, adding an ORDER BY expression to the resulting query.
+func (q *Query) OrderBy(sql string) *Query {
+	clone := q.clone()
+
+	clone.orderByStatements = make([]string, len(q.orderByStatements)+1)
+	copy(clone.orderByStatements, q.orderByStatements)
+	clone.orderByStatements[len(q.orderByStatements)] = sql
+
+	return clone
+}
+
+// Offset clones the query object, adding an OFFSET expression to the resulting query.
+// The given integer parameter is added to the set of arguments for use with the final query.
+func (q *Query) Offset(offset int) *Query {
+	clone := q.clone()
+	clone.offset = strconv.Itoa(offset)
+
+	return clone
+}
+
+// Limit clones the query object, adding a LIMIT expression to the resulting query.
+// The given integer parameter is added to the set of arguments for use with the final query.
+func (q *Query) Limit(limit int) *Query {
+	clone := q.clone()
+	clone.limit = strconv.Itoa(limit)
+
+	return clone
+}
+
+// Bind clones the query object, adding or replacing the given key and value as an argument to the
+// resulting query.
+// String and integer values are supported, and will be exploded automatically into a list of
+// arguments.
+func (q *Query) Bind(key string, value interface{}) *Query {
+	// It's a natural mistake to write the key prefixed with the colon: fix that and move on.
+	key = strings.TrimLeft(key, ":")
+
+	clone := q.clone()
+	clone.args = make(map[string]interface{})
+	for k, v := range q.args {
+		clone.args[k] = v
+	}
+	clone.args[key] = value
+
+	return clone
+}
+
+// String generates the final query for execution.
+func (q *Query) String() string {
+	var query string
+
+	if len(q.selectStatements) > 0 {
+		query += "SELECT " + strings.Join(q.selectStatements, ", ")
+	}
+	if len(q.fromStatements) > 0 {
+		query += " FROM " + strings.Join(q.fromStatements, ", ")
+	}
+	if len(q.joinStatements) > 0 {
+		query += " " + strings.Join(q.joinStatements, " ")
+	}
+	if len(q.whereStatements) > 0 {
+		query += " WHERE " + strings.Join(q.whereStatements, " AND ")
+	}
+	if len(q.orderByStatements) > 0 {
+		query += " ORDER BY " + strings.Join(q.orderByStatements, ", ")
+	}
+	if len(q.limit) > 0 {
+		query += " LIMIT " + q.limit
+	}
+	if len(q.offset) > 0 {
+		query += " OFFSET " + q.offset
+	}
+
+	// If an argument named :Ids is bound to an string or integer slice or array, explode in
+	// place as :Ids_0, :Ids_1, ... :Ids_N.
+	for key, value := range q.args {
+		rt := reflect.TypeOf(value)
+		switch rt.Kind() {
+		case reflect.Slice:
+			fallthrough
+		case reflect.Array:
+			switch rt.Elem().Kind() {
+			case reflect.String:
+				fallthrough
+			case reflect.Int:
+				valueLen := reflect.ValueOf(value).Len()
+
+				keys := []string{}
```

Comment:
No objection here, though binding an empty array is almost an error, e.g. `WHERE X IN (:EmptyArray)` will resolve to `WHERE X IN ()` at the moment. There's way to avoid this, but you need to expose a bigger surface API, e.g. `WhereIn` to intelligently remove the condition.